prusti-specs/src/common.rs

```diff
@@ -1,29 +1,338 @@
+            rewriter.rewrite_impl_item_method(self);
+        }
+    }
+
+    struct Rewriter<'a> {
+        self_type: &'a TypePath,
+        self_type_trait: Option<&'a TypePath>,
+    }
+
+    impl<'a> Rewriter<'a> {
+        pub fn rewrite_impl_item_method(&mut self, item: &mut ImplItemMethod) {
+            syn::visit_mut::visit_impl_item_method_mut(self, item);
+        }
+
+        pub fn rewrite_item_fn(&mut self, item: &mut syn::ItemFn) {
+            syn::visit_mut::visit_item_fn_mut(self, item);
+        }
+    }
+
+    impl<'a> VisitMut for Rewriter<'a> {
+        fn visit_type_path_mut(&mut self, ty_path: &mut syn::TypePath) {
+            if ty_path.qself.is_none()
+                && !ty_path.path.segments.is_empty()
+                && ty_path.path.segments[0].ident == "Self" {
+                if ty_path.path.segments.len() == 1 {
+                    // replace `Self` type
+                    *ty_path = self.self_type.clone();
+                } else if ty_path.path.segments.len() >= 2 {
+                    // replace associated types
+                    let mut path_rest = ty_path.path.segments.clone()
+                        .into_pairs()
+                        .skip(1)
+                        .collect::<syn::punctuated::Punctuated<syn::PathSegment, _>>();
+                    if ty_path.path.segments.trailing_punct() {
+                        path_rest.push_punct(<syn::Token![::]>::default());
+                    }
+                    let self_type = &self.self_type;
+                    let self_type_trait = &self.self_type_trait;
+                    let new_type_path: syn::TypePath = parse_quote_spanned! {ty_path.span()=>
+                    < #self_type as #self_type_trait > :: #path_rest
+                };
+                    *ty_path = new_type_path;
+                }
+            }
+            syn::visit_mut::visit_type_path_mut(self, ty_path);
+        }
     }
 }
 
-impl HasGenerics for syn::ItemStruct {
-    fn get_generics(&self) -> &syn::Generics {
-        &self.generics
+/// See [RewritableReceiver]
+mod receiver_rewriter {
+    use proc_macro2::{Ident, TokenStream, TokenTree};
+    use quote::{quote, quote_spanned, ToTokens};
+    use syn::{FnArg, ImplItemMethod, ItemFn, Macro, parse_quote_spanned};
+    use syn::spanned::Spanned;
+    use syn::visit_mut::VisitMut;
+
+    /// Rewrites the receiver of a method-like item.
+    /// This can be used to convert impl methods to free-standing functions.
+    ///
+    /// # Example
+    /// For a provided new type `T`, the function
+    /// ```ignore
+    /// fn foo(&mut self, arg1: Bar) -> Baz {
+    ///     self.qux()
+    /// }
+    /// ```
+    /// will be rewritten to
+    /// ```ignore
+    /// fn foo(_self: &mut T, arg1: Bar) -> Baz {
+    ///     _self.qux()
+    /// }
+    /// ```
+    pub(crate) trait RewritableReceiver {
+        fn rewrite_receiver<T: ToTokens>(&mut self, new_ty: &T);
```

Comment:
Can the `T` here be something specific, e.g. `TypePath`? Is there a use for providing different tokenisable types?

Reply:
I think the original idea of this was because the implementation is purely token based, i.e. `TypePath` would be tokenized anyway. But I think this is a good idea, the interface should be designed according to client's needs, not the implementation.

Reply:
And actually, I think this was not considered during my refactoring (of the refactoring). `T` makes no sense anymore.